File: floodns/src/main/java/ch/ethz/systems/floodns/ext/utils/EdgeColoring.java
package ch.ethz.systems.floodns.ext.utils;

import org.apache.commons.lang3.tuple.ImmutablePair;

import java.util.*;

public class EdgeColoring {

    public static Map<Integer, Integer> colorEdges(Map<Integer, ImmutablePair<Integer, Integer>> commodities) {
        HashSet<String> nodes = new HashSet<>();
        for (ImmutablePair<Integer, Integer> conn : commodities.values()) {
            nodes.add(conn.getKey() + "-");
            nodes.add(conn.getValue() + "+");
        }
        HashMap<String, Integer> nodes_mapping = new HashMap<>();
        int idx = 0;
        for (String node : nodes) {
            nodes_mapping.put(node, idx);
            idx++;
        }

        List<ImmutablePair<Integer, Integer>> edges = new ArrayList<>();

        for (ImmutablePair<Integer, Integer> conn : commodities.values()) {
            String src = conn.getKey() + "-";
            String dst = conn.getValue() + "+";
            edges.add(new ImmutablePair<>(nodes_mapping.get(src), nodes_mapping.get(dst)));
        }

        int l = nodes.size();
        int r = nodes.size();
        int m = edges.size();
        int[] A = new int[m];
        int[] B = new int[m];

        for (int i = 0; i < m; i++) {
            A[i] = edges.get(i).getKey();
            B[i] = edges.get(i).getValue();
        }

        RegularGraph ret = buildKRegularGraph(l, r, A, B);
        int K = ret.getK();
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> ord = new ArrayList<>();
        for (int i = 0; i < A.length; i++) {
            ord.add(i);
        }

        rec(ret.getC(), ret.getD(), ord, K, ans, ret.getP());
        int[] colors = new int[m];

        for (int i = 0; i < ans.size(); i++) {
            for (int j : ans.get(i)) {
                if (j < m) {
                    colors[j] = i;
                }
            }
        }

        Map<Integer, Integer> assignments = new HashMap<>();

        for (Map.Entry<Integer, ImmutablePair<Integer, Integer>> entry : commodities.entrySet()) {
            assignments.put(entry.getKey(), colors[entry.getValue().getKey()]);
        }

        return assignments;
    }

    private static RegularGraph buildKRegularGraph(int n, int m, int[] A, int[] B) {
        int[] dega = new int[n];
        int[] degb = new int[m];

        for (int a : A) {
            dega[a]++;
        }

        for (int b : B) {
            degb[b]++;
        }

        int K = Math.max(Arrays.stream(dega).max().getAsInt(), Arrays.stream(degb).max().getAsInt());
        UnionFind UFa = contract(dega, K);
        int[] ida = new int[n];
        int pa = 0;

        for (int i = 0; i < n; i++) {
            if (UFa.find(i) == i) {
                ida[i] = pa;
                pa++;
            }
        }

        UnionFind UFb = contract(degb, K);
        int[] idb = new int[m];
        int pb = 0;

        for (int i = 0; i < m; i++) {
            if (UFb.find(i) == i) {
                idb[i] = pb;
                pb++;
            }
        }

        int p = Math.max(pa, pb);
        dega = new int[p];
        degb = new int[p];
        List<Integer> C = new ArrayList<>();
        List<Integer> D = new ArrayList<>();

        for (int i = 0; i < A.length; i++) {
            int u = ida[UFa.find(A[i])];
            int v = idb[UFb.find(B[i])];
            C.add(u);
            D.add(v);
            dega[u]++;
            degb[v]++;
        }

        int j = 0;

        for (int i = 0; i < p; i++) {
            while (dega[i] < K) {
                while (degb[j] == K) {
                    j++;
                }

                C.add(i);
                D.add(j);
                dega[i]++;
                degb[j]++;
            }
        }

        return new RegularGraph(K, p, C, D);
    }

    private static UnionFind contract(int[] deg, int k) {
        PriorityQueue<int[]> hq = new PriorityQueue<>((a, b) -> a[0] - b[0]);

        for (int i = 0; i < deg.length; i++) {
            hq.add(new int[]{deg[i], i});
        }

        UnionFind UF = new UnionFind(deg.length);

        while (hq.size() >= 2) {
            int[] p = hq.poll();
            int[] q = hq.poll();

            if (p[0] + q[0] > k) {
                continue;
            }

            p[0] += q[0];
            UF.union(p[1], q[1]);
            hq.add(p);
        }

        return UF;
    }

    private static List<Integer> eulerTrail(List<Integer> A, List<Integer> B, List<Integer> ord, int n) {
        int V = 2 * n;
        List<List<ImmutablePair<Integer, Integer>>> G = new ArrayList<>();
        for (int i = 0; i < V; i++) {
            G.add(new ArrayList<>());
        }

        int m = 0;
        for (int i : ord) {
            G.get(A.get(i)).add(new ImmutablePair<>(B.get(i) + n, m));
            G.get(B.get(i) + n).add(new ImmutablePair<>(A.get(i), m));
            m++;
        }

        boolean[] used_v = new boolean[V];
        boolean[] used_e = new boolean[m];
        List<Integer> ans = new ArrayList<>();

        for (int i = 0; i < V; i++) {
            if (used_v[i]) {
                continue;
            }

            Stack<ImmutablePair<Integer, Integer>> st = new Stack<>();
            List<Integer> ord2 = new ArrayList<>();
            st.push(new ImmutablePair<>(i, -1));

            while (!st.isEmpty()) {
                ImmutablePair<Integer, Integer> p = st.peek();
                int id_ = p.getKey();
                used_v[id_] = true;

                if (G.get(id_).isEmpty()) {
                    ord2.add(p.getValue());
                    st.pop();
                } else {
                    ImmutablePair<Integer, Integer> e = G.get(id_).get(G.get(id_).size() - 1);
                    G.get(id_).remove(G.get(id_).size() - 1);

                    if (used_e[e.getValue()]) {
                        continue;
                    }

                    used_e[e.getValue()] = true;
                    st.push(e);
                }
            }

            ord2.remove(ord2.size() - 1);
            Collections.reverse(ord2);
            ans.addAll(ord2);
        }

        for (int i = 0; i < ans.size(); i++) {
            ans.set(i, ord.get(ans.get(i)));
        }

        return ans;
    }

    private static void rec(List<Integer> A, List<Integer> B, List<Integer> ord, int K, List<List<Integer>> ans, int n) {
        if (K == 0) {
            return;
        } else if (K == 1) {
            ans.add(ord);
            return;
        } else if ((K & 1) == 1) {
            HopcroftKarp G = new HopcroftKarp(n, n);

            for (int i : ord) {
                G.addEdges(A.get(i), B.get(i));
            }

            G.maxMatching();
            List<Integer> lst = new ArrayList<>();
            ans.add(new ArrayList<>());

            for (int i : ord) {
                if (G.getMatchLeft()[A.get(i)] == B.get(i)) {
                    G.getMatchLeft()[A.get(i)] = -1;
                    ans.get(ans.size() - 1).add(i);
                } else {
                    lst.add(i);
                }
            }

            rec(A, B, lst, K - 1, ans, n);
        } else {
            List<Integer> path = eulerTrail(A, B, ord, n);
            List<Integer> L = new ArrayList<>();
            List<Integer> R = new ArrayList<>();

            for (int i = 0; i < path.size(); i++) {
                if ((i & 1) == 1) {
                    L.add(path.get(i));
                } else {
                    R.add(path.get(i));
                }
            }

            rec(A, B, L, K / 2, ans, n);
            rec(A, B, R, K / 2, ans, n);
        }
    }
}

File: floodns/external/edge_coloring/edge_coloring.py
from collections import deque
from heapq import *


class HopCroftKarp:
    def __init__(self, n, m):
        self.n = n
        self.m = m
        self.G = [[] for _ in range(n)]
        self.RG = [[] for _ in range(m)]
        self.match_l = [-1] * n
        self.match_r = [-1] * m
        self.used = [0] * n
        self.time_stamp = 0

    def add_edges(self, u, v):
        self.G[u].append(v)

    def _build_argument_path(self):
        queue = deque()
        self.dist = [-1] * self.n
        for i in range(self.n):
            if self.match_l[i] == -1:
                queue.append(i)
                self.dist[i] = 0
        while queue:
            a = queue.popleft()
            for b in self.G[a]:
                c = self.match_r[b]
                if c >= 0 and self.dist[c] == -1:
                    self.dist[c] = self.dist[a] + 1
                    queue.append(c)

    def _find_min_dist_argument_path(self, a):
        self.used[a] = self.time_stamp
        for b in self.G[a]:
            c = self.match_r[b]
            if c < 0 or (
                self.used[c] != self.time_stamp
                and self.dist[c] == self.dist[a] + 1
                and self._find_min_dist_argument_path(c)
            ):
                self.match_r[b] = a
                self.match_l[a] = b
                return True
        return False

    def max_matching(self):
        while 1:
            self._build_argument_path()
            self.time_stamp += 1
            flow = 0
            for i in range(self.n):
                if self.match_l[i] == -1:
                    flow += self._find_min_dist_argument_path(i)
            if flow == 0:
                break
        ret = []
        for i in range(self.n):
            if self.match_l[i] >= 0:
                ret.append((i, self.match_l[i]))
        return ret


class UnionFind:
    def __init__(self, n):
        self.n = n
        self.parents = [-1] * n
        self.group = n

    def find(self, x):
        if self.parents[x] < 0:
            return x
        else:
            self.parents[x] = self.find(self.parents[x])
            return self.parents[x]

    def union(self, x, y):
        x = self.find(x)
        y = self.find(y)

        if x == y:
            return
        self.group -= 1
        if self.parents[x] > self.parents[y]:
            x, y = y, x

        self.parents[x] += self.parents[y]
        self.parents[y] = x

    def size(self, x):
        return -self.parents[self.find(x)]

    def same(self, x, y):
        return self.find(x) == self.find(y)

    def members(self, x):
        root = self.find(x)
        return [i for i in range(self.n) if self.find(i) == root]

    def roots(self):
        return [i for i, x in enumerate(self.parents) if x < 0]

    def group_count(self):
        return self.group

    def all_group_members(self):
        dic = {r: [] for r in self.roots()}
        for i in range(self.n):
            dic[self.find(i)].append(i)
        return dic

    def __str__(self):
        return "\n".join("{}: {}".format(r, self.members(r)) for r in self.roots())


def contract(deg, k):
    hq = []
    for i, d in enumerate(deg):
        hq.append([d, i])
    heapify(hq)
    UF = UnionFind(len(deg))
    while len(hq) >= 2:
        p = heappop(hq)
        q = heappop(hq)
        if p[0] + q[0] > k:
            continue
        p[0] += q[0]
        UF.union(p[1], q[1])
        heappush(hq, p)
    return UF


def build_k_regular_graph(n, m, A, B):
    dega = [0] * n
    degb = [0] * m
    for a in A:
        dega[a] += 1
    for b in B:
        degb[b] += 1
    K = max(*dega, *degb)

    UFa = contract(dega, K)
    ida = [-1] * n
    pa = 0
    for i in range(n):
        if UFa.find(i) == i:
            ida[i] = pa
            pa += 1

    UFb = contract(degb, K)
    idb = [-1] * m
    pb = 0
    for i in range(m):
        if UFb.find(i) == i:
            idb[i] = pb
            pb += 1

    p = max(pa, pb)
    dega = [0] * p
    degb = [0] * p

    C = []
    D = []
    for i in range(len(A)):
        u = ida[UFa.find(A[i])]
        v = idb[UFb.find(B[i])]
        C.append(u)
        D.append(v)
        dega[u] += 1
        degb[v] += 1
    j = 0
    for i in range(p):
        while dega[i] < K:
            while degb[j] == K:
                j += 1
            C.append(i)
            D.append(j)
            dega[i] += 1
            degb[j] += 1

    return K, p, C, D


def EdgeColoring(a, b, A, B):
    K, n, A, B = build_k_regular_graph(a, b, A, B)

    ord = [i for i in range(len(A))]
    ans = []

    def euler_trail(ord):
        V = 2 * n
        G = [[] for _ in range(V)]
        m = 0
        for i in ord:
            G[A[i]].append((B[i] + n, m))
            G[B[i] + n].append((A[i], m))
            m += 1
        used_v = [0] * V
        used_e = [0] * m
        ans = []
        for i in range(V):
            if used_v[i]:
                continue
            st = []
            ord2 = []
            st.append((i, -1))
            while st:
                id_ = st[-1][0]
                used_v[id_] = True
                if len(G[id_]) == 0:
                    ord2.append(st[-1][1])
                    st.pop()
                else:
                    e = G[id_][-1]
                    G[id_].pop()
                    if used_e[e[1]]:
                        continue
                    used_e[e[1]] = True
                    st.append(e)
            ord2.pop()
            ord2 = ord2[::-1]
            ans += ord2
        for i, a in enumerate(ans):
            ans[i] = ord[a]
        return ans

    def rec(ord, K):
        if K == 0:
            return
        elif K == 1:
            ans.append(ord)
            return
        elif K & 1:
            G = HopCroftKarp(n, n)
            for i in ord:
                G.add_edges(A[i], B[i])
            G.max_matching()
            lst = []
            ans.append([])
            for i in ord:
                if G.match_l[A[i]] == B[i]:
                    G.match_l[A[i]] = -1
                    ans[-1].append(i)
                else:
                    lst.append(i)
            rec(lst, K - 1)
        else:
            path = euler_trail(ord)
            L = []
            R = []
            for i, p in enumerate(path):
                if i & 1:
                    L.append(p)
                else:
                    R.append(p)
            rec(L, K // 2)
            rec(R, K // 2)

    rec(ord, K)
    return K, ans


def color_edges(commodities: dict):
    nodes = set()
    for src, dst in commodities.values():
        nodes.add(f"{src}-")
        nodes.add(f"{dst}+")
    nodes_mapping = {node: i for i, node in enumerate(nodes)}
    edges = [
        (nodes_mapping[f"{src}-"], nodes_mapping[f"{dst}+"]) for src, dst in commodities.values()
    ]

    l = r = len(nodes)
    m = len(edges)

    A = []
    B = []
    for src, dst in edges:
        A.append(src)
        B.append(dst)

    K, ans = EdgeColoring(l, r, A, B)
    colors = [-1] * m
    for i in range(len(ans)):
        for j in ans[i]:
            if j < m:
                colors[j] = i
    assignments = {}
    for conn_id, color in zip(commodities.keys(), colors):
        assignments[conn_id] = color

    return assignments
